rust/user-info-fetcher/src/main.rs

```diff
@@ -0,0 +1,262 @@
+        .serve(app.into_make_service())
+        .with_graceful_shutdown(shutdown_requested)
+        .await
+        .context(RunServerSnafu)
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase", untagged)]
+enum UserInfoRequest {
+    UserInfoRequestById(UserInfoRequestById),
+    UserInfoRequestByName(UserInfoRequestByName),
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase")]
+struct UserInfoRequestById {
+    id: String,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Eq, Hash, Clone)]
+#[serde(rename_all = "camelCase")]
+struct UserInfoRequestByName {
+    username: String,
+}
+
+#[derive(Serialize, Clone, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+struct UserInfo {
+    /// This might be null in case the id is not known (e.g. the backend does not have this info).
+    id: Option<String>,
+    /// This might be null in case the username is not known (e.g. the backend does not have this info).
+    username: Option<String>,
+    groups: Vec<String>,
+    custom_attributes: HashMap<String, Vec<String>>,
+}
+
+#[derive(Snafu, Debug)]
+#[snafu(module)]
+enum GetUserInfoError {
+    #[snafu(display("failed to get user information from Keycloak"))]
+    Keycloak { source: backend::keycloak::Error },
+}
+
+impl http_error::Error for GetUserInfoError {
+    fn status_code(&self) -> hyper::StatusCode {
+        // todo: the warn here loses context about the scope in which the error occurred, eg: stackable_opa_user_info_fetcher::backend::keycloak
+        // Also, we should make the log level (warn vs error) more dynamic in the backend's impl `http_error::Error for Error`
+        tracing::warn!(error = ?self, "Error while processing request");
+        match self {
+            Self::Keycloak { source } => source.status_code(),
+        }
+    }
+}
+
+async fn get_user_info(
+    State(state): State<AppState>,
+    Json(req): Json<UserInfoRequest>,
+) -> Result<Json<UserInfo>, http_error::JsonResponse<Arc<GetUserInfoError>>> {
+    let AppState {
+        config,
+        http,
+        credentials,
+        user_info_cache,
+    } = state;
+    Ok(Json(
+        user_info_cache
+            .try_get_with_by_ref(&req, async {
+                match &config.backend {
+                    crd::Backend::None {} => {
+                        let user_id = match &req {
+                            UserInfoRequest::UserInfoRequestById(UserInfoRequestById { id }) => {
+                                Some(id.clone())
+                            }
+                            _ => None,
+                        };
+                        let user_name = match &req {
+                            UserInfoRequest::UserInfoRequestByName(UserInfoRequestByName {
+                                username,
+                            }) => Some(username.clone()),
+                            _ => None,
+                        };
```

Comment:
Nit: useless clone

```suggestion
                        let (user_id, username) = match req {
                            UserInfoRequest::UserInfoRequestById(UserInfoRequestById { id }) => {
                                (Some(id), None)
                            }
                            UserInfoRequest::UserInfoRequestByName(UserInfoRequestByName {
                                username,
                            }) => (None, Some(username))
                        };
```

Reply:
Fixed in f1b533222c8f5da331c392c342a6a969a105de43

Reply:
Actually reverting that because...

```
 252 | ...                   id: user_id,
    |                           ^^^^^^^ expected `Option<String>`, found `Option<&String>`
```